Basic input class

Korijn opened the pull request:
Closes #664

Adds a basic Input class with an example demonstrating how to implement simple FPS controls with it

kushalkolar commented:
You guys are epic and so fast! With pointer events I can already see how it can simplify our [drag PR](https://github.com/fastplotlib/fastplotlib/pull/413) ! I didn't realize we can just subclass from the existing Controller :D

Korijn commented:
I'm working to at least include all the basic functionality in this PR with pointer state as well.

almarklein commented on pygfx/controllers/_input.py:
```diff
@@ -0,0 +1,35 @@
+
+class Input:
+    def __init__(self, register_events=None):
+        self.key_down = {}
+        self.key_down = {}
```
Duplicate line.

Korijn commented:
This is now working. Could one of you give it a try locally and review?

@almarklein I was thinking to move the FPS controls into a controller class, but it will have a different architecture and design than the other controllers. I also don't want to refactor all of them. Do you have any ideas on how to deal with that? I was thinking about creating another subpackage `pygfx.input`...

Korijn commented on examples/feature_demo/input_fps.py:
```diff
@@ -0,0 +1,129 @@
+        movement -= (
+            np.dot(movement, camera.local.reference_up) * camera.local.reference_up
+        )
+    camera.local.position = camera.local.position + movement
```
@FirefoxMetzger I just wanted you to see how elegant this FPS controls implementation is thanks to your work <3

almarklein replied:
Impressive with how few lines of code this takes! ❤️

Korijn commented on pygfx/controllers/_time.py:
```diff
@@ -0,0 +1,40 @@
+
+    @property
+    def delta(self):
+        """Returns the time in seconds since the last frame."""
```
This is actually the time the previous frame took to render 🤦‍♂️

almarklein replied:
It's likely even longer than the time it took to render the frame (fps limiting).

Would we want to perhaps smooth this value, to avoid variations in it from affecting the way the interaction feels?

Korijn replied:
After thinking about this some more, I think it's correctly implemented:

* For every `after_flush` event, there is only 1 call to `perf_counter()`, so we don't "lose time"
* The `delta` should have the same value for all update handlers running in a given frame
* The `delta` should be equal to the time since the previous point in time where update handlers were executed, which is usually right after `after_flush`

almarklein commented on pygfx/controllers/_input.py:
```diff
@@ -0,0 +1,121 @@
+        elif type == "wheel":
+            self._wheel_delta[event.pointer_id] = event.dy
+
+    def key_down(self, key):
+        """Returns True if the key_down event was received in the current frame"""
+        return self._key_down.get(key.lower(), False)
```
Is there a benefit to this, compared to listening for key_down events directly? Also in theory a user can press a key twice, and using this method only see one. Or is that intentional?

Korijn replied:
It's an abstraction that makes it possible for many once-per-frame update functions to avoid the boilerplate of receiving events and tracking state of inputs :)

almarklein replied:
I see the convenience, but it also add the constraint that any code that uses this must run "in sync", i.e. exactly once per draw.

Korijn replied:
Yes, although that is intentional

Korijn replied:
I also wanted to mention, in an earlier state of this code I had used `input.key(...)` instead of `input.key_down(...)`. It was very tricky to toggle fly mode because pressing the key would make `input.key(...)` evaluate to `True` for multiple frames. In other words, it would rapidly toggle where I just wanted to toggle it once. That's the use case for `input.key_down(...)`.

almarklein commented on examples/feature_demo/input_fps.py:
```diff
@@ -0,0 +1,129 @@
+renderer = gfx.renderers.WgpuRenderer(canvas)
+scene = gfx.Group()
+
+# hide the cursor, lock it to the window and keep it centered
+glfw.set_input_mode(canvas._window, glfw.CURSOR, glfw.CURSOR_DISABLED)
```
Nice! Can we also listen to Esc to free the mouse again?

kushalkolar replied:
Just played with it and I agree :laughing:

Korijn replied:
Haha, how often do you play shooter video games? Usually you alt tab out to get the mouse back. But yes, I can try to implement that.

kushalkolar replied:
Just tried and alt+tab works! my gaming muscle memory is usually not present when working and I just panic instead :P

Korijn replied:
Added some code to release the mouse with escape (and ignore mouse events while it is released). Clicking on the window captures it again.

almarklein commented:
Tried it and it works really nice!  Also feels immersive because the mouse is grabbed.

>  I was thinking about creating another subpackage pygfx.input...

Yes, I think that makes the most sense. I think controllers are quite different because they closely relate to cameras. Perhaps we can (later) refactor the controllers to make use of the input class(es) to make them simpler. I think wrapping makes more sense than subclassing here, what do you think?

kushalkolar commented:
The example works well for me along with small modifications to try out various things (moving the torus and the view, sensitivities, etc.) :smile:

Korijn commented:
* Added escape/mouse click handlers to release/capture the mouse
* Added acceleration to make movement smoother

The example is "done" I think, next step is reorganizing the code into a `pygfx.inputs` folder.

Korijn commented:
The more I think about this the more I feel like it should be a part of wgpu-py's GUI layer... I will see what that feels like by trying it out on a branch.

kushalkolar commented:
I'll give this a shot in wgpu if you can hint me at where it would go!

Korijn commented:
> I'll give this a shot in wgpu if you can hint me at where it would go!

`wgpu/gui/input.py` :)

almarklein commented:
Closing. We will get back to this when we work on #495. We see two approaches, but we'll have to see at the time:
* The animation event has info on what buttons are down.
* The canvas has API to check what buttons are down.
